### PR TITLE
Bug in DataFrame.replace casts columns to ``object`` dtype if items in ``to_replace`` not in values

### LINKED ISSUE #32988: REGR: `replace` casts columns to `object`
#### Calling `df.replace` casts columns to `object`

```python
pd.DataFrame(np.eye(2)).replace(to_replace=[None, -np.inf, np.inf], value=pd.NA).dtypes
# 0    object
# 1     object
# dtype: object
```
#### Problem description
I'd expect that the dtypes of the columns remain the same after replacing with pd.NA (especially considering no values are replaced in the above call). We do not get this issue if `to_replace` is any subset of `[None, -np.inf, np.inf]`. We get the same issue if `value` is instead `np.nan`.

#### Expected Output
```python
# 0    float64
# 1     float64
# dtype: object
```
#### Output of ``pd.show_versions()``

<details>
Installed with `pip install -e .` on current master. Was also able to reproduce on `1.0.1`.

INSTALLED VERSIONS
------------------
commit           : bed9103e52ae412774c22b8e198f7b66708981c7
python           : 3.7.0.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
Version          : Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.26.0.dev0+2701.gbed9103e5
numpy            : 1.18.2
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.1.post20200323
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by simonjayhawkins**
- [ ] closes #32988
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
why are you reverting things?

**Comment by jreback** [comment]
I am ok with reverting things for the 1.0.4 branch but not for master.

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -905,7 +874,6 @@ def putmask(
 
         # if we are passed a scalar None, convert it here
         if not is_list_like(new) and isna(new) and not self.is_object:
-            # FIXME: make sure we have compatible NA
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:908]: pls leave this and other comments

Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py]: sure

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @simonjayhawkins

**Pull request merged**